Few fix on the AST

hogoww opened the pull request:
hope I didn't forget to commit something (working on compiler related stuff crashes a bit !)

Fix for: #3250. Added an intermediary hook that retains retro compatibility. There still may be retro compatibility problems for people that overrode the #visitTemporaryNodes: in a way that mimic the previous behavior.

added fix for #3262 as well. retro compatibility should be fine, by default behavior.
This should disappear at some point, given that it might be miss used by user.

fix for: #3237 Moved the code out in a visitor. Retains both the #dump selector & the GT hook in RBProgramNode though.
I didn't rename it either because I didn't follow all the conversation between steph' and vince.
Added some small comments, that @aranega should at least check :)

For a separation of the concern in the issues, added a #visitLiteralNode: AND a #visitLiteralValueNode: so it work with both version. (see issue #3262)

hogoww committed:
[fix) issue #3250

hogoww committed:
[fix] delocalisation of dump for RBAST

hogoww committed:
[comments] [ cleanup] removed an unused method, and add some comments related to dump.

hogoww committed:
[retrocompatibility]  added an override for visitLiteralValueNode, which doesn't exists as long as PR #3261 isn't accepted

hogoww committed:
[retrocompatibility]  added an override for visitLiteralValueNode, which doesn't exists as long as PR #3261 isn't accepted

Ducasse commented:
Thanks. 
For me dump is a really bad name, asLiteralExpressions would be a lot clearer. 
This is not because people around are using strange that we should follow them.

hogoww commented:
Agreed, but I can't do it myself :p

Should I port those on Pharo 7 as well?

Ducasse commented on src/AST-Core-Traits/TRBProgramNodeVisitor.trait.st:
```diff
@@ -128,6 +128,11 @@ TRBProgramNodeVisitor >> visitSuperNode: aSuperNode [
 
+{ #category : #visiting }
+TRBProgramNodeVisitor >> visitTemporaryDeclarationNode: aTemporaryDeclarationNode [
+	^self visitArgumentNode: aTemporaryDeclarationNode 
```
can you add a comment stating that this is a backward compatible solution because this is strange that visitArgumentNode: visit a temp.

hogoww replied:
Took the one I put in RBVisitor, as suggested in the your last comment

Ducasse commented on src/AST-Core/RBDumpVisitor.class.st:
```diff
@@ -0,0 +1,214 @@
+"
+I represent a visitor that generates Pharo code
```
I'm a visitor that generates code whose execution will recreate the visited node (similarly to storeOn: protocol). 
This is handy because we can simply serialize the object in a textual form without requiring a separate parser. 

```
(RBParser parseExpression: (Point >> #asNonFractionalPoint) ast dump ) formattedCode 

'RBMethodNode
	selector: #asNonFractionalPoint
	body:
		(RBSequenceNode
			statements:
				{(RBMessageNode
					receiver:
						(RBMessageNode
							receiver:
								(RBMessageNode
									receiver: (RBInstanceVariableNode named: #x)
									selector: #isFraction)
							selector: #or:
							arguments:
								{(RBBlockNode
									body:
										(RBSequenceNode
											statements:
												{(RBMessageNode
													receiver: (RBInstanceVariableNode named: #y)
													selector: #isFraction)}))})
					selector: #ifTrue:
					arguments:
						{(RBBlockNode
							body:
								(RBSequenceNode
									statements:
										{(RBReturnNode
											value:
												(RBMessageNode
													receiver:
														(RBMessageNode
															receiver: (RBInstanceVariableNode named: #x)
															selector: #asFloat)
													selector: #@
													arguments:
														{(RBMessageNode
															receiver: (RBInstanceVariableNode named: #y)
															selector: #asFloat)}))}))})})'
```

hogoww replied:
Thanks, that's indeed better than what I wrote !

Ducasse commented on src/AST-Core/RBLiteralNode.class.st:
```diff
@@ -28,6 +28,11 @@ RBLiteralNode >> = anObject [
 
+{ #category : #visiting }
+RBLiteralNode >> acceptVisitor: aProgramNodeVisitor [
+	^aProgramNodeVisitor visitLiteralNode:self
```
Please use the code formatter
^ aProgramNodeVisitor visitLiteralNode: self

hogoww replied:
Done

Ducasse commented on src/AST-Core/RBLiteralValueNode.class.st:
```diff
@@ -46,22 +46,14 @@ RBLiteralValueNode >> = anObject [
 { #category : #visiting }
 RBLiteralValueNode >> acceptVisitor: aProgramNodeVisitor [ 
-	^aProgramNodeVisitor visitLiteralNode: self
+	^aProgramNodeVisitor visitLiteralValueNode: self
```
^ aProgramNodeVisitor visitLiteralValueNode: self

hogoww replied:
Done

Ducasse commented on src/AST-Core/RBProgramNode.class.st:
```diff
@@ -293,12 +293,11 @@ RBProgramNode >> do: aBlock [
-{ #category : #generation }
-RBProgramNode >> dumpOn: aStream [
-	self subclassResponsibility 
+	"Generate the Pharo code from self"
```
Generate a literal expression that recreates the receiver

hogoww replied:
replaced.

Ducasse commented on src/AST-Core/RBProgramNodeVisitor.class.st:
```diff
@@ -77,6 +77,7 @@ RBProgramNodeVisitor >> visitMethodNode: aMethodNode [
 	aMethodNode pragmas
 		do: [ :each | self visitNode: each ].
 	self visitNode: aMethodNode body
+
```
remove empty line

hogoww replied:
done

Ducasse commented on src/AST-Core/RBProgramNodeVisitor.class.st:
```diff
@@ -127,6 +128,14 @@ RBProgramNodeVisitor >> visitSuperNode: aSuperNode [
 ]
 
+{ #category : #visiting }
+RBProgramNodeVisitor >> visitTemporaryDeclarationNode: aTemporaryDeclarationNode [ 
```
Can you use this comment for the method in the trait

hogoww replied:
done

hogoww committed:
[fix] requested fix for PR + format on every method of RBProgramNodeVisitor

hogoww committed:
Merge pull request #1 from hogoww/bugCommit

[fix] requested fix for PR + format on every method of RBProgramNodeV…

hogoww committed:
[fix] Format #acceptVisitor: selector for RBNodes, for consistency.

hogoww committed:
Merge pull request #2 from hogoww/bugCommit

[fix] Format #acceptVisitor: selector for RBNodes, for consistency.

hogoww committed:
Update TRBProgramNodeVisitor.trait.st

unimportant change (although the comment is indeed incomplete) to try to restart Travis, just in case.

hogoww committed:
Update TRBProgramNodeVisitor.trait.st

try to restart travis, attempt number 2

hogoww commented:
Tried to investigate the failling build, but i didn't get it. Sorry.

hogoww committed:
[fix] Corrected SHRBTextStyler, to work with the 'new' way of visiting literals

hogoww committed:
[fix] Add LiteralValueNode visit on TRBPRgramNodeVisitor + comments to explain the behavior

hogoww commented:
> Tried to investigate the failling build, but i didn't get it. Sorry.
It was a random problem that happened, build seems to be fine now.
I hope i didn't add/remove anything stupid

MarcusDenker commented:
close and re-open tu run latest tests

MarcusDenker reviewed (approved):
ok
